db/impl/etcd/client.go

```diff
@@ -317,3 +318,141 @@ func (c *Client) ListPrefix(prefix string, obj db.Entity) ([]db.Entity, error) {
 
 	return c.traverse(resp.Node, obj), nil
 }
+
+// Acquire and permanently hold a lock. Attempts until timeout. If timeout is
+// zero, it will only try once.
+func (c *Client) Acquire(lock db.Lock) error {
+	logrus.Debugf("Acquiring lock %v", lock)
+
+	content, err := jsonio.Write(lock)
+	if err != nil {
+		return errors.LockFailed.Combine(err)
+	}
+
+	path, err := lock.Path()
+	if err != nil {
+		return errors.LockFailed.Combine(err)
+	}
+
+	_, err = c.client.Set(context.Background(), c.qualified(path), string(content), &client.SetOptions{PrevExist: client.PrevNoExist})
+	if er, ok := err.(client.Error); ok && er.Code == 105 {
+		if lock.MayExist() {
+			_, err := c.client.Set(context.Background(), c.qualified(path), string(content), &client.SetOptions{PrevExist: client.PrevExist, PrevValue: string(content)})
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}
+	} else if err != nil {
+		return err
+	}
+
+	logrus.Debugf("Acquired lock %v", lock)
+
+	return nil
+}
+
+// Free a lock. Pass force=true to force it dead.
+func (c *Client) Free(lock db.Lock, force bool) error {
+	content, err := jsonio.Write(lock)
+	if err != nil {
+		return errors.LockFailed.Combine(err)
+	}
+
+	path, err := lock.Path()
+	if err != nil {
+		return errors.LockFailed.Combine(err)
+	}
+
+	opts := &client.DeleteOptions{PrevValue: string(content)}
+
+	if force {
+		opts = nil
+	}
+
+	_, err = c.client.Delete(context.Background(), c.qualified(path), opts)
+	return err
+}
+
+// AcquireAndRefresh starts a goroutine to refresh the key every 1/4
+// (jittered) of the TTL. A stop channel is returned which, when sent a
+// struct, will terminate the refresh. Error is returned for anything that
+// might occur while setting up the goroutine.
+//
+// Do not use Free to free these locks, it will not work! Use the stop
+// channel.
+func (c *Client) AcquireAndRefresh(lock db.Lock, ttl time.Duration) (chan struct{}, error) {
+	logrus.Debugf("In refresh, performing preliminary permanent lock on %v", lock)
+	if err := c.Acquire(lock); err != nil {
+		return nil, err
+	}
+
+	stopChan := make(chan struct{}, 1)
+
+	go func() {
+		for {
+			select {
+			case <-stopChan:
+				if err := c.Free(lock, false); err != nil {
+					logrus.Errorf("Error freeing lock %q: %v", lock, err)
+				}
+
+				return
+			default:
+				time.Sleep(wait.Jitter(ttl/4, 0))
+				if err := c.AcquireWithTTL(lock, ttl); err != nil {
+					logrus.Errorf("Could not acquire lock %v: %v", lock, err)
+				}
+			}
+		}
+	}()
+
+	return stopChan, nil
+}
+
+// AcquireWithTTL acquires a lock with a TTL by using two compare-and-swap
+// operations to refresh any lock that exists by us, and to take any lock that
+// is not taken.
+func (c *Client) AcquireWithTTL(lock db.Lock, ttl time.Duration) error {
+	if ttl < 0 {
```

Comment:
TTL 0 clears the TTL, but we should actually make a minimum; etcd gets 
really unreliable when you start writing under 250ms/hit.

On 24 Aug 2016, at 9:32, Bill Robinson wrote:

>> +				time.Sleep(wait.Jitter(ttl/4, 0))
>> +				if err := c.AcquireWithTTL(lock, ttl); err != nil {
>> +					logrus.Errorf("Could not acquire lock %v: %v", lock, err)
>> +				}
>> +			}
>> +		}
>> +	}()
>> +
>> +	return stopChan, nil
>> +}
>> +
>> +// AcquireWithTTL acquires a lock with a TTL by using two 
>> compare-and-swap
>> +// operations to refresh any lock that exists by us, and to take any 
>> lock that
>> +// is not taken.
>> +func (c *Client) AcquireWithTTL(lock db.Lock, ttl time.Duration) 
>> error {
>> +	if ttl < 0 {
>
> Is there a use case for a 0 TTL lock?  I guess you could use it to 
> check for the presence of a lock...
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/contiv/volplugin/pull/444/files/2b3726db59c59297cc9d671a722b2fb161d35f37#r76091743